R/derive_param_tte.R

```diff
@@ -0,0 +1,413 @@
+#'   added.
+#'
+#'   \item The selected observations of all event source datasets are combined into a
+#'   single dataset.
+#'
+#'   \item For each patient the first observation (with respect to the `ADT`
+#'   variable) from the single dataset is selected. }
+#'
+#'   **Deriving the censoring observations:**
+#'
+#'   \enumerate{ \item For each censoring source dataset the observations as
+#'   specified by the `filter` element are selected. Then for each patient the
+#'   last observation (with respect to `date`) is selected.
+#'
+#'   \item The `ADT` variable is set to the variable specified by the
+#'   \code{date} element. If the date variable is a datetime variable, only
+#'   the datepart is copied. If the source variable is a character variable, it
+#'   is converted to a date. If the date is incomplete, it is imputed as the
+#'   first possible date.
+#'
+#'   \item The `CNSR` variable is added and set to the \code{censor} element.
+#'
+#'   \item The variables specified by the \code{set_values_to} element are
+#'   added.
+#'
+#'   \item The selected observations of all censoring source datasets are
+#'   combined into a single dataset.
+#'
+#'   \item For each patient the last observation (with respect to the `ADT`
+#'   variable) from the single dataset is selected. }
+#'
+#'   For each subject (as defined by the `subject_keys` parameter) an
+#'   observation is selected. If an event is available, the event observation is
+#'   selected. Otherwise the censoring observation is selected.
+#'
+#'   Finally
+#'   1. the variables specified for `start_date` and `start_imputation_flag` are
+#'   joined from the ADSL dataset,
+#'   1. the variables as defined by the `set_values_to` parameter are added,
+#'   1. the `ADT`/`ADTM` variable is set to the maximum of `ADT`/`ADTM` and
+#'   `STARTDT`/`STARTDTM` (depending on the `create_datetime` parameter), and
+#'   1. the new observations are added to the output dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @return The input dataset with the new parameter added
+#'
+#' @keywords derivation adtte
+#'
+#' @export
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("adsl")
+#'
+#' death <- tte_source(
+#'   dataset = adsl,
+#'   filter = DTHFL == "Y",
+#'   date = DTHDT,
+#'   set_values_to =vars(
+#'     EVENTDESC = "DEATH",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "DTHDT"))
+#'
+#' lstalv <- tte_source(
+#'   dataset = adsl,
+#'   date = LSTALVDT,
+#'   censor = 1,
+#'   set_values_to = vars(
+#'     EVENTDESC = "LAST KNOWN ALIVE DATE",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "LSTALVDT"))
+#'
+#' derive_param_tte(
+#'   dataset_adsl = adsl,
+#'   event_conditions = list(death),
+#'   censor_conditions = list(lstalv),
+#'   set_values_to = vars(PARAMCD = "OS",
+#'                        PARAM = "Overall Survival")) %>%
+#' select(-STUDYID) %>% `[`(20:30,)
+derive_param_tte <- function(dataset = NULL,
```

Comment:
In the [ADTTE workflow](https://exbpbox.box.com/s/bvrbn2ie94nk3di3z5t52v2zgi9q9o1l) `AVAL` is derived in the second step because usually the derivation of `AVAL` is the same across all parameters. Are you aware of scenarios where this could not work?

Reply:
Thanks for sharing the workflow document, and I agree AVAL is usually the same derivation. Though I've seen parameters used in sensitivity analysis, where discounts are added to different time intervals between the start date and ADT when deriving AVAL. This may be too specific, and users can probably address it during post-processing if they want to overwrite the standard AVAL derivation for a certain parameter?

Reply:
Yes, I think in this scenario `mutate()`, `case_when()`, and `compute_duration()` could be used in step 2 of the ADTTE workflow.